tests/unit/test_charm.py

```diff
@@ -80,14 +80,42 @@ def test_given_private_key_is_stored_and_certificate_relation_is_created_when_on
             },
         )
         patch_generate_csr.return_value = csr.encode()
+        self.harness._backend.secret_add(label="csr", content={"csr": csr})
         self.harness.add_relation(relation_name="certificates", remote_app="certificates-provider")
 
         self.harness.update_config(key_values={"subject": "banana.com"})
 
         patch_request_certificate.assert_called_with(certificate_signing_request=csr.encode())
 
+    @patch(
+        "charms.tls_certificates_interface.v2.tls_certificates.TLSCertificatesRequiresV2.request_certificate_creation"  # noqa: E501, W505
+    )
+    @patch("charm.generate_csr")
+    def test_given_private_key_is_stored_and_csr_is_not_stored_certificate_relation_is_created_when_on_config_changed_then_certificate_is_not_requested(  # noqa: E501
```

Comment:
We could add one more test case that validate generating the csr.
maybe something like:
`test_given_csr_not_stored_when_config_changed_then_csr_is_generated()`

Reply:
The test  "def test_given_certificate_relation_is_not_created_when_on_config_changed_then_csr_is_generated" was also validating the same functionality. I renamed it as   "def test_given_certificate_relation_is_not_created_csr_is_not_stored_when_on_config_changed_then_csr_is_generated" to explicitly show that csr is not stored at the beginning.
Also added  another test "def test_given_certificate_relation_is_created_csr_is_not_stored_when_on_config_changed_then_csr_is_generated".